### PR TITLE
Simplify decoder impl by making it object oriented, avoid pointers where possible

### PR DESCRIPTION
**Pull request by reuben**

### CODE REVIEW

**Comment by reuben** [comment]
I would recommend looking at this commit by commit.

File: native_client/ctcdecode/ctc_beam_search_decoder.h [hunk 1]
```diff
@@ -7,66 +7,67 @@
...
- *     ext_scorer: External scorer to evaluate a prefix, which consists of
- *                 n-gram language model scoring and word insertion term.
- *                 Default null, decoding the input sample without scorer.
- * Return:
- *     A struct containing prefixes and state variables. 
-*/
-DecoderState* decoder_init(const Alphabet &alphabet,
-                           int class_dim,
-                           Scorer *ext_scorer);
+  Scorer* ext_scorer_; // weak
```
Comment (lissyx) [inline native_client/ctcdecode/ctc_beam_search_decoder.h]: shouldn't we use `std::weak_ptr` then ?

Comment (reuben) [reply native_client/ctcdecode/ctc_beam_search_decoder.h]: `std::weak_ptr` is the weak counterpart of `std::shared_ptr`. The corresponding owned pointer for this pointer is a `std::unique_ptr`, which has no weak counterpart. Most people use a bare pointer. There's a proposal to add a `std::observer_ptr`: http://open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3840.pdf

File: native_client/alphabet.h [hunk 2]
```diff
@@ -15,7 +15,14 @@
  */
 class Alphabet {
 public:
+  Alphabet() {
+  }
```
Comment (lissyx) [inline native_client/alphabet.h]: nit: `Alphabet() {}`

File: native_client/ctcdecode/scorer.cpp [hunk 3]
```diff
@@ -30,18 +30,16 @@ static const int32_t MAGIC = 'TRIE';
...
+               double beta)
+  : is_character_based_(true)
   , max_order_(0)
-  , alphabet_(alphabet)
 {
   reset_params(alpha, beta);
+}
 
+void
+Scorer::init(const std::string& lm_path, const std::string& trie_path)
```
Comment (lissyx) [inline native_client/ctcdecode/scorer.cpp]: We pass `lm_path` and `trie_path` on `init` as well as on the constructor (below) ?

File: native_client/ctcdecode/scorer.cpp [hunk 4]
```diff
@@ -56,13 +54,26 @@ Scorer::Scorer(double alpha,
   setup(lm_path, trie_path);
 }
 
+Scorer::Scorer(double alpha,
+               double beta,
+               const std::string& lm_path,
+               const std::string& trie_path,
```
Comment (lissyx) [inline native_client/ctcdecode/scorer.cpp]: We pass `lm_path` and `trie_path` on the constructor as well as on `init` (above) ?

Comment (lissyx) [reply native_client/ctcdecode/scorer.cpp]: Would it make sense to make them members of `Scorer` like `alphabet_` ?

File: native_client/alphabet.h [hunk 5]
```diff
@@ -15,15 +15,15 @@
  */
 class Alphabet {
 public:
-  Alphabet() {
-  }
-
-  Alphabet(const char *config_file) {
-    init(config_file);
-  }
+  Alphabet() = default;
```
Comment (lissyx) [inline native_client/alphabet.h]: Ok, never mind my previous comment on that

File: native_client/generate_trie.cpp [hunk 6]
```diff
@@ -8,7 +8,10 @@
 using namespace std;
 
 int generate_trie(const char* alphabet_path, const char* kenlm_path, const char* trie_path) {
-  Alphabet alphabet(alphabet_path);
+  Alphabet alphabet;
+  if (int err = alphabet.init(alphabet_path)) {
```
Comment (lissyx) [inline native_client/generate_trie.cpp]: feels like a death trap :-), maybe we should just:
```
int err = alphabet.init(alphabet_path);
if (err != 0) {
  return err;
}
```

Comment (lissyx) [reply native_client/generate_trie.cpp]: or just do the same than we do in `modelstate.cc`

File: native_client/trie_load.cc [hunk 7]
```diff
@@ -16,7 +16,10 @@ int main(int argc, char** argv)
 
   printf("Loading trie(%s) and alphabet(%s)\n", trie_path, alphabet_path);
 
-  Alphabet alphabet(alphabet_path);
+  Alphabet alphabet;
```
Comment (lissyx) [inline native_client/trie_load.cc]: Same, maybe we should just have the same code as in `modelstate.cc`

File: native_client/ctcdecode/scorer.cpp (hunk 4 above)
Comment (reuben) [reply native_client/ctcdecode/scorer.cpp]: I don't understand the suggestion. What would making them members accomplish?

File: native_client/generate_trie.cpp (hunk 6 above)
Comment (reuben) [reply native_client/generate_trie.cpp]: `modelstate.cc` does the same as is done here. By death trap do you mean the code style with the declaration inside the condition? I can change that.

File: native_client/ctcdecode/scorer.cpp (hunk 4 above)
Comment (lissyx) [reply native_client/ctcdecode/scorer.cpp]: Is there any reason we have two entry points to set those vlaues ?

File: native_client/generate_trie.cpp (hunk 6 above)
Comment (reuben) [reply native_client/generate_trie.cpp]: (Although I don't see the death trappiness of it)

File: native_client/ctcdecode/scorer.cpp (hunk 4 above)
Comment (reuben) [reply native_client/ctcdecode/scorer.cpp]: We don't. `init` is private.

File: native_client/generate_trie.cpp (hunk 6 above)
Comment (lissyx) [reply native_client/generate_trie.cpp]: Yes, this is what I mean. Well, I always find that construction tricky, and error-prone in future evolutions.

File: native_client/ctcdecode/scorer.cpp (hunk 4 above)
Comment (lissyx) [reply native_client/ctcdecode/scorer.cpp]: Right, then I think it would be simpler to just have `lm_path_ = lm_path` and have `init()` using them

Comment (reuben) [reply native_client/ctcdecode/scorer.cpp]: They aren't used anywhere outside of setup_lm. There's no reason to inflate the object size.

Comment (lissyx) [reply native_client/ctcdecode/scorer.cpp]: r=me then

Comment (reuben) [reply native_client/ctcdecode/scorer.cpp]: I've merged init and setup to avoid this problem.

File: native_client/ctcdecode/scorer.cpp [hunk 8]
```diff
@@ -61,16 +53,26 @@ Scorer::Scorer(double alpha,
...
                const std::string& trie_path,
                const std::string& alphabet_config_path)
-  : Scorer(alpha, beta, lm_path, trie_path, Alphabet(alphabet_config_path.c_str()))
-{
-}
-
-Scorer::~Scorer()
+  : Scorer(alpha, beta)
 {
+  alphabet_.init(alphabet_config_path.c_str());
```
Comment (dabinat) [inline native_client/ctcdecode/scorer.cpp]: Looks like we're missing a check on the return value of init here.

**Pull request merged**